Mousefood display trait

deadbaed opened the pull request:
This is the main pull request of #66, introducing a trait for display backends to implement to work with mousefood. This will also remove the `flush_callback` function, and removes the requirement of using a framebuffer.

deadbaed commented:
Done! Those changes build and run fine on the simulator and my ssd1306 display.

I just bought a ST7789, I'll be able to test with it as soon as it arrives.

j-g00da commented on mousefood/src/backend.rs:
```diff
@@ -14,14 +12,17 @@ use ratatui_core::backend::{Backend, ClearType};
 
-/// Embedded backend configuration.
-pub struct EmbeddedBackendConfig<D, C>
+pub trait MousefoodDisplay<D, C>
```
nit: I would move this to a separate file/module.

j-g00da reviewed (commented):
Please run fmt and clippy and then fix clippy warnings.
Also the framebuffer tests are panicking after changes.
I will take a deeper look later.

deadbaed commented:
The git log is messy, but I wanted to do everything I wanted functionality wise before requesting review. I'll rebase and re-make the commits later, make other PRs if needed.

- Renamed the trait to `DisplayTarget`, moved it to its own module `display`
- removed as many trait parameters as possible (maybe more could be removed?)
- refactored usage of `TermColor(Color::Reset, TermColorType::Background)`
- started to add minimal documentation to make clippy happy, but they will need to be improved -- to be frank I am really not the best at those kind of things, I want to improve though.

I will start to look into proc macros and derive macros, to see what's possible.

j-g00da commented on mousefood/src/backend.rs:
```diff
@@ -15,13 +14,7 @@ use ratatui_core::layout;
-    C: PixelColor,
-{
-    /// Callback fired after each buffer flush.
-    pub flush_callback: Box<dyn FnMut(&mut D)>,
```
Just a note: Removing flush_callback makes this a breaking change

j-g00da commented on mousefood/src/backend.rs:
```diff
@@ -107,30 +90,33 @@ where
-    ) -> EmbeddedBackend<'display, D, C> {
+        display: &'display mut M,
+        config: EmbeddedBackendConfig,
+    ) -> Result<EmbeddedBackend<'display, D, C, M>> {
```
This is also a breaking change which must be documented

j-g00da commented on mousefood/src/error.rs:
```diff
@@ -1,12 +1,20 @@
 //! Mousefood `Error` enum.
 
+pub(crate) type Result<T, E = Error> = core::result::Result<T, E>;
```
I think we can make this public

j-g00da commented on mousefood/src/display.rs:
```diff
@@ -0,0 +1,24 @@
+/// Trait to implement for displays to be usable with mousefood.
+///
+/// The display must implement [`DrawTarget`]
+pub trait DisplayTarget<D, C>
```
I'm not sold on a name, maybe `BufferedDisplay`? It tells more about what this is, and doesn't get mixed up with `DrawTarget`.

j-g00da commented on mousefood/src/display.rs:
```diff
@@ -0,0 +1,24 @@
+    ///
+    /// If the display driver requires additional operations, this is the place to make them
+    fn flush(&mut self) -> Result<()>;
+}
```
One more thing, we should think about is making the integration crates able to define custom color conversions. E.g. Weact integration has it's own colors which ideally should be moved out of the main crate.
Let me know if you have some ideas for it.

deadbaed replied:
I think integration crates should be able to define custom color conversions (the Weact case confirms us that).

Is taking anything implementing `embedded_graphics::PixelColor` enough? Mabye it's time for another trait 😄

deadbaed replied:
Discovered `DrawTargetExt` from embedded_graphics, with a function to convert colors, would this interest you? https://docs.rs/embedded-graphics/latest/embedded_graphics/draw_target/trait.DrawTargetExt.html#tymethod.color_converted

j-g00da replied:
Sorry for late response, I'm not sure if this can be helpful here but you can experiment with this a bit.

j-g00da commented on mousefood/src/colors.rs:
```diff
@@ -11,6 +11,13 @@ pub enum TermColorType {
 
 pub struct TermColor(pub Color, pub TermColorType);
 
+impl TermColor {
+    /// Initial color to use when initializing buffered and unbuffered displays
+    pub const fn initial_color() -> Self {
+        TermColor(Color::Reset, TermColorType::Background)
+    }
+}
```
Maybe let's just use a `Default` trait for that? I think it would make sense as a default value.

j-g00da commented:
> Something about implementing DrawTarget for future Mousefood Displays: since some displays may have hardware acceleration, we would need to implement all functions of DrawTarget, in order to keep the potential acceleration.

But we are calling it directly on underlying display using `.draw_target()` anyway, so it shouldn't matter. We shouldn't need to implement `DrawTarget` at all.

deadbaed commented on examples/simulator/src/main.rs:
```diff
@@ -27,31 +27,64 @@ use ratatui::widgets::{Block, Paragraph, Wrap};
+        }
+    }
+
+    impl mousefood::embedded_graphics::prelude::DrawTarget for MySuperDisplay {
+        type Color = Bgr565;
+
+        type Error = core::convert::Infallible;
+
+        fn draw_iter<I>(&mut self, pixels: I) -> Result<(), Self::Error>
+        where
+            I: IntoIterator<Item = mousefood::embedded_graphics::Pixel<Self::Color>>,
+        {
+            self.display.draw_iter(pixels)
+        }
+    }
+
+    impl mousefood::embedded_graphics::prelude::Dimensions for MySuperDisplay {
+        fn bounding_box(&self) -> mousefood::embedded_graphics::primitives::Rectangle {
+            self.display.bounding_box()
+        }
+    }
```
Hmm, I think at the start of the trait experiment I was requiring structs implementing the trait to also implement `DrawTarget`, but it is not!

I'm relieved I can remove that

j-g00da commented on mousefood/src/prelude.rs:
```diff
@@ -7,7 +7,7 @@
 
 pub use crate::backend::{EmbeddedBackend, EmbeddedBackendConfig};
-pub use crate::display::DisplayTarget;
+pub use crate::display::BufferedDisplay;
```
I don't think we need this in prelude. This will be mostly used by integration crates and rarely by users developing actual apps.

deadbaed commented:
For the 2 breaking changes, do you want one commit per breaking change, or having both changes in the same commit is fine?

j-g00da commented:
One is fine

deadbaed commented on mousefood/src/error.rs:
```diff
@@ -6,7 +6,13 @@ pub enum Error {
+    /// Selected [`ClearType`](ratatui_core::backend::ClearType) is not supported by Mousefood.
     #[error("ClearType::{0} is not supported by Mousefood")]
     ClearTypeUnsupported(alloc::string::String),
+    /// Flushing to the display failed.
+    #[error("flushing to DrawTarget failed")]
+    Flush,
```
I am debating whether this should be a standalone structure, because currently the only usage of `Error::Flush` is outside of `mousefood`, which is a bit weird imo.

I will make a new structure `FlushError`, export it, and have a `From<FlushError> for Error` to be able to use our `Result` type.

deadbaed replied:
I'll revert to just taking a string, it will be simpler, and it's a good starting point.

deadbaed commented:
Ready for review! The only thing missing is good documentation, if you have any ideas I'm all ears!

j-g00da commented:
Thanks! Let's move out these two to new PRs and merge these first.

deadbaed commented on mousefood/src/lib.rs:
```diff
@@ -4,15 +4,18 @@
 mod default_font;
 pub mod error;
-mod framebuffer;
+pub mod framebuffer;
```
Works for me! But that would mean displays without built-in buffer would need to implement it themselves (if they track mousefood from the main branch)

deadbaed replied:
Absolutely, I'll will make that PR! Didn't properly read your last comment about making PRs before merging this one...

deadbaed commented on mousefood/src/lib.rs:
```diff
@@ -4,15 +4,18 @@
 pub use backend::{EmbeddedBackend, EmbeddedBackendConfig};
 pub use embedded_graphics;
 
+pub type Result<T, E = error::Error> = core::result::Result<T, E>;
```
Works for me

deadbaed replied:
After some thoughts, export of this type alias is no longer needed for this PR. If you want I can still export it in another PR.

j-g00da replied:
Ok, let's leave it private then. 
If we decide we want to use it in integration crates, then we can make it public in the future.

deadbaed commented on mousefood/src/buffered_display.rs:
```diff
@@ -0,0 +1,28 @@
+pub use embedded_graphics::prelude::DrawTarget;
+use embedded_graphics::prelude::PixelColor;
+
+/// Flushing to the display failed.
+#[derive(thiserror::Error, Debug)]
+#[error("flushing to DrawTarget failed: {0}")]
+pub struct FlushError(pub alloc::string::String);
```
I think the additional string is nice to provide more detail (it is used in the simulator to tell that the window got closed).

I can make it optional, so that downstream consumers can provide additional detail on why the flushing failed.

For reference my ssd1306 returns [this error](https://docs.rs/display-interface/latest/display_interface/enum.DisplayError.html), we can pass it like this:
```rust
        fn flush(&mut self) -> Result<(), mousefood::buffered_display::FlushError> {
            self.display
                .flush()
                .map_err(|e| mousefood::buffered_display::FlushError(format!("{e:?}")))
        }
```


The ST7789 display return [this error](https://docs.rs/mipidsi/latest/mipidsi/interface/enum.SpiError.html), we can pass it like this:
```rust
        fn flush(&mut self) -> Result<(), FlushError> {
            use mousefood::embedded_graphics::draw_target::DrawTarget;
            use mousefood::embedded_graphics::prelude::Dimensions;

            let rectangle = self.draw_target().bounding_box();
            self.display
                .fill_contiguous(&rectangle, &self.framebuffer)
                .map_err(|e| mousefood::buffered_display::FlushError(format!("{e:?}")))
        }
```

j-g00da commented:
I think we should implement crates for simulator and weact drivers in this PR already.

j-g00da commented:
Let's use this branch as staging, just don't rebase anymore so we can safely work on this together.

j-g00da commented on mousefood/src/backend.rs:
```diff
@@ -55,18 +43,14 @@ where
+    B: BufferedDisplay<D, C>,
     D: DrawTarget<Color = C> + 'display,
-    C: PixelColor + 'display,
+    C: PixelColor + 'display + From<TermColor>,
```
Btw. we should probably use `Into<PixelColor>` instead of `PixelColor + From<TermColor>`, which doesn't make much sense and will make it hard to define color conversions outside the main mousefood crate.

j-g00da commented:
So I experimented a bit with what has to be done to be able to define custom color conversions without issues.
First we have to remove `From<TermColor>` bounds from `C` type params.
Then to make it work we can create a static method in `BufferedDisplay` like this (similar to what `DrawTargetExt` provides):
```rust
fn convert_color(color: TermColor) -> C;
```

And use it in `Backend` like this:
```rust
let mut style_builder = MonoTextStyleBuilder::new()
    .font(&self.font_regular)
    .text_color(B::convert_color(TermColor(cell.fg, TermColorType::Foreground)))
    .background_color(B::convert_color(TermColor(cell.bg, TermColorType::Background)));
```
(this can be simplified, but that's the general idea)

Finally we can implement conversions inside integration crates (example for epd-waveshare):
```rust
impl<SPI, BUSY, DC, RST, DELAY> BufferedDisplay<Display4in2, epd_waveshare::color::Color>
for WaveshareEpd4in2<SPI, BUSY, DC, RST, DELAY>
where
    SPI: SpiDevice,
    BUSY: InputPin,
    DC: OutputPin,
    RST: OutputPin,
    DELAY: DelayNs,
{
    fn draw_target(&mut self) -> &mut impl DrawTarget<Color=Color> {
        todo!()
    }

    fn flush(&mut self) -> Result<(), FlushError> {
        todo!()
    }

    fn convert_color(color: TermColor) -> epd_waveshare::color::Color {
        // TermColor -> BinaryColor (implemented in mousefood)
        let binary_color: BinaryColor = color.into();
        // BinaryColor -> Color (from epd_waveshare)
        binary_color.into()
    }
}
```

Let me know what you think @deadbaed

deadbaed commented:
Awesome, just what I thought. I'd love to test your changes on my `ssd1306`, if you have something you're happy with, feel free to make a PR!

I did some shopping and bought a `ILI9341` display (thinking I was buying a `ST7789` like you have), and a WeAct red/black/white display. When I'll have more time I'll get them up running with mousefood, and use them as test targets for integration crates.
